src/state/liveTokens/liveTokensStore.ts

```diff
@@ -0,0 +1,271 @@
+  metadata: {
+    currency: string;
+    requestId: string;
+    requestTime: string;
+  };
+  result: LiveTokensData;
+  errors?: string[];
+};
+
+type LiveTokensParams = {
+  subscribedTokensByRoute: TokenSubscriptionCountByRoute;
+  activeRoute: string;
+  currency: SupportedCurrencyKey;
+};
+
+type UpdateSubscribedTokensParams = {
+  route: string;
+  tokenIds: string[];
+};
+
+type LiveTokenStoreState = {
+  subscribedTokensByRoute: TokenSubscriptionCountByRoute;
+  tokens: LiveTokensData;
+};
+
+type LiveTokenStoreActions = {
+  removeSubscribedTokens: ({ route, tokenIds }: UpdateSubscribedTokensParams) => void;
+  addSubscribedTokens: ({ route, tokenIds }: UpdateSubscribedTokensParams) => void;
+  clear: () => void;
+};
+
+type LiveTokensStore = LiveTokenStoreState & LiveTokenStoreActions;
+
+const initialState: LiveTokenStoreState = {
+  subscribedTokensByRoute: {},
+  tokens: {},
+};
+
+const fetchTokensData = async ({ subscribedTokensByRoute, activeRoute, currency }: LiveTokensParams): Promise<LiveTokensData | null> => {
+  const tokenIds = Object.keys(subscribedTokensByRoute[activeRoute] || {}).map(tokenId => {
+    if (tokenId.includes('_')) {
+      return convertLegacyTokenIdToTokenId(tokenId);
+    }
+    return tokenId;
+  });
+
+  if (tokenIds.length === 0) {
+    return null;
+  }
+
+  // Separate ETH variants from other tokens
+  const ethVariants: string[] = [];
+  const otherTokens: string[] = [];
+
+  tokenIds.forEach(tokenId => {
+    if (isEthVariant(tokenId)) {
+      ethVariants.push(tokenId);
+    } else {
+      otherTokens.push(tokenId);
+    }
+  });
+
+  const tokensToFetch = [
+    ...otherTokens,
+    // Only subscribe to mainnet ETH if we have any ETH variants
+    ...(ethVariants.length > 0 ? [ETH_MAINNET_TOKEN_ID] : []),
+  ];
+
+  const response = await getPlatformClient().get<LiveTokensResponse>('/prices/GetCurrentPrices', {
+    params: {
+      tokenIds: tokensToFetch.join(','),
+      currency,
+    },
+  });
+
+  const result: LiveTokensData = {};
+
+  Object.entries(response.data.result).forEach(([tokenId, tokenData]) => {
+    if (tokenId !== ETH_MAINNET_TOKEN_ID) {
+      result[convertTokenIdToLegacyTokenId(tokenId)] = tokenData;
+    }
+  });
+
+  // Map ETH data to all ETH variants
+  if (response.data.result[ETH_MAINNET_TOKEN_ID]) {
+    const ethData = response.data.result[ETH_MAINNET_TOKEN_ID];
+    ethVariants.forEach(ethVariant => {
+      result[convertTokenIdToLegacyTokenId(ethVariant)] = ethData;
+    });
+  }
+
+  return result;
+};
+
+function updateUserAssetsStore(tokens: LiveTokensData) {
+  useUserAssetsStore.getState().updateTokens(tokens);
+}
+
+export const useLiveTokensStore = createQueryStore<LiveTokensData | null, LiveTokensParams, LiveTokensStore>(
+  {
+    fetcher: fetchTokensData,
+    disableCache: true,
+    staleTime: time.seconds(5),
+    setData: ({ data, set }) => {
+      if (data) {
+        set(state => {
+          updateUserAssetsStore(data);
```

Comment:
Since `updateUserAssetsStore` has no dependency on `state`, don't see any reason to call it from within `set`. Doesn't make any actual difference since `setData` is already within `set` call, but I think the intent is clearer if it's called before or after.

Probably most appropriate to call it from `onFetched`, which runs after `setData` and outside of any `set` calls.